src/osd/OSD.cc

```diff
@@ -8805,6 +8816,54 @@ void OSD::handle_op(OpRequestRef& op, OSDMapRef& osdmap)
   }
 }
 
+void OSD::handle_backoff(OpRequestRef& op, OSDMapRef& osdmap)
+{
+  MOSDBackoff *m = static_cast<MOSDBackoff*>(op->get_req());
+  Session *s = static_cast<Session*>(m->get_connection()->get_priv());
+  dout(10) << __func__ << " " << *m << " session " << s << dendl;
+  assert(s);
+  s->put();
+
+  if (m->op != CEPH_OSD_BACKOFF_OP_ACK_BLOCK) {
+    dout(10) << __func__ << " unrecognized op, ignoring" << dendl;
+    return;
+  }
+
+  // map hobject range to PG(s)
+  bool queued = false;
+  hobject_t pos = m->begin;
+  while (true) {
+    pg_t _pgid(pos.get_hash(), pos.pool);
+    if (osdmap->have_pg_pool(pos.pool)) {
+      _pgid = osdmap->raw_pg_to_pg(_pgid);
+    }
+    spg_t pgid;
+    if (!osdmap->get_primary_shard(_pgid, &pgid)) {
+      // missing pool or acting set empty -- drop
```

Comment:
Is this a normal situation? If the backoff spans across multiple PGs, this could break the loop in the middle before ops to all PGs are dispatched.